source/compiler-core/slang-tint-compiler.cpp

```diff
@@ -0,0 +1,162 @@
+        virtual SLANG_NO_THROW bool SLANG_MCALL isFileBased() SLANG_OVERRIDE
+        {
+            return false;
+        }
+
+        virtual SLANG_NO_THROW SlangResult SLANG_MCALL getVersionString(
+            slang::IBlob** outVersionString) SLANG_OVERRIDE;
+
+        SlangResult compile(IArtifact *const sourceArtifact, IArtifact** outArtifact);
+
+        SlangResult init(ISlangSharedLibrary* library);
+
+    protected:
+
+        ComPtr<ISlangSharedLibrary> m_sharedLibrary;
+
+    private:
+
+        tint_CompileFunc m_compile;
+        tint_FreeBufferFunc m_freeBuffer;
+    };
+
+    SlangResult TintDownstreamCompiler::init(ISlangSharedLibrary* library)
+    {
+        tint_CompileFunc compile =
+            (tint_CompileFunc)library->findFuncByName("tint_compile");
+        if (compile == nullptr)
+        {
+            return SLANG_FAIL;
+        }
+
+        tint_FreeBufferFunc freeBuffer =
+            (tint_FreeBufferFunc)library->findFuncByName("tint_free_buffer");
+        if (m_freeBuffer == nullptr)
+        {
+            return SLANG_FAIL;
+        }
+
+        m_sharedLibrary = library;
+        m_desc = Desc(SLANG_PASS_THROUGH_TINT);
+        m_compile = compile;
+        m_freeBuffer = freeBuffer;
+        return SLANG_OK;
+    }
+
+    SlangResult TintDownstreamCompilerUtil::locateCompilers(
+        const String& path,
+        ISlangSharedLibraryLoader* loader,
+        DownstreamCompilerSet* set)
+    {
+        ComPtr<ISlangSharedLibrary> library;
+        SLANG_RETURN_ON_FAIL(
+            DownstreamCompilerUtil::loadSharedLibrary(
+                path, loader, nullptr, "slang-tint", library));
+        SLANG_ASSERT(library);
+
+        TintDownstreamCompiler *const compiler = new TintDownstreamCompiler();
+        SLANG_RETURN_ON_FAIL(compiler->init(library));
+
+        set->addCompiler(ComPtr<IDownstreamCompiler>(compiler));
+        return SLANG_OK;
+    }
+
+    SlangResult TintDownstreamCompiler::compile(
+        const CompileOptions& options, IArtifact** outArtifact)
+    {
+        IArtifact *const sourceArtifact = options.sourceArtifacts[0];
+        return compile(sourceArtifact, outArtifact);
+    }
+
+    SlangResult TintDownstreamCompiler::compile(
+        IArtifact *const sourceArtifact, IArtifact** outArtifact)
+    {
+        tint_CompileRequest req = {};
+
+        if (sourceArtifact == nullptr)
+            return SLANG_FAIL;
+
+        ComPtr<ISlangBlob> sourceBlob;
+        SLANG_RETURN_FALSE_ON_FAIL(
+            sourceArtifact->loadBlob(ArtifactKeep::Yes, sourceBlob.writeRef())
+        );
+
+        std::string
+            wgslCode(
+                (char*)sourceBlob->getBufferPointer(),
+                (char*)sourceBlob->getBufferPointer() + sourceBlob->getBufferSize());
+        req.wgslCode = wgslCode.c_str();
+        req.wgslCodeLength = wgslCode.size();
+
+        tint_CompileResult result = {};
+        int const resultCode {m_compile(&req, &result)};
+        if (resultCode != 0)
```

Comment:
Leaking memory on tint_compileResult.

Reply:
Use SLANG_DEFER to ensure things are freed:

auto rs = m_compile(&result);
SLANG_DEFER(m_free(&result));
if (rs != 0) ...

Reply:
Do you mean m_compile fails? A common convention (which is followed in this case) is to make sure that no resources are actually returned if the call fails.

    extern "C"
    SLANG_SHARED_LIB_EXPORT
    int tint_compile(tint_CompileRequest* request, tint_CompileResult* result)
    {
        ///...
        uint8_t *const resultBuffer {static_cast<uint8_t*>(::malloc(resultBufferSize))};
        if(resultBuffer == nullptr)
            return 1;
        ::memcpy(resultBuffer, spirvCode.data(), resultBufferSize);
    
        result->buffer = resultBuffer;
        result->bufferSize = resultBufferSize;
        return 0;
    }

Reply:
(But of course I may as well add a DEFER anyway.)

Reply:
How do we report back diagnostics returned by tint?

Reply:
Done. I added an error field to the slang-tint result struct.


https://github.com/shader-slang/dawn/blob/slang-tint/src/slang-tint/slang-tint.cpp#L62